### PR TITLE
Prioritized active managers by total active managers config and priority assigned to each manager

### PR DESCRIPTION
**Pull request by keitabyte**
Prioritized active managers by total active managers config and priority assigned to each manager

Summary of PR
- The ability to set a manager as active or inactive, where an active manager will actively process messages while an inactive manager does not.
- An introduction of `PrioritizedActiveManager` which allows setting a priority to a given manager along with a configurable total of active managers. With each individual manager's heartbeat, the set of available managers that are online will strive to have the highest priority managers active up to the the total number of configured total active managers.
- Re-queue stale in-progress queues' messages back into its origin queue, which can occur when resizing the number of pods of managers. With this fix, we can now confidently resize the number of pods assigned to a given queue without manual intervention of verifying/migrating stale in-progress messages.

### CODE REVIEW

File: fetcher.go [hunk 1]
```diff
@@ -13,8 +14,11 @@ import (
 //Fetcher is an interface for managing work messages
 type Fetcher interface {
 	Queue() string
+	InProgressQueue() string
 	Fetch()
 	Acknowledge(*Msg)
+	Active(bool)
```
Comment (stefannegrea) [inline fetcher.go:20]: Can we please rename this to `SetActive`, I think it would be a better name.

File: storage/redis.go [hunk 2]
```diff
@@ -59,52 +59,133 @@ func (r *redisStore) CheckRtt(ctx context.Context) int64 {
 
 var sidekiqHeartbeatJobsKey = ":work"
 
-func (r *redisStore) SendHeartbeat(ctx context.Context, heartbeat *Heartbeat) error {
+func (r *redisStore) SendHeartbeat(ctx context.Context, heartbeat *Heartbeat, heartbeatManagerTTL time.Duration) error {
```
Comment (stefannegrea) [inline storage/redis.go:62]: I would make this TTL part of the `Heatbeat` struct and extract the value just like the other values below.

File: storage/storage.go [hunk 3]
```diff
@@ -73,9 +88,24 @@ type Store interface {
 	GetAllStats(ctx context.Context, queues []string) (*Stats, error)
 
 	// Heartbeat
-	SendHeartbeat(ctx context.Context, heartbeat *Heartbeat) error
-	RemoveHeartbeat(ctx context.Context, heartbeat *Heartbeat) error
```
Comment (stefannegrea) [inline storage/storage.go:77]: I like the symmetry of this interface, any reason to not keep the interface as-is?

File: storage/storage.go [hunk 4]
```diff
@@ -73,9 +88,24 @@ type Store interface {
...
+	// Task runner expiration operations
+	GetExpiredTaskRunnerKeys(ctx context.Context, expireTS int64) ([]string, error)
+	GetTaskRunnerInfo(ctx context.Context, taskRunnerKey string) (TaskRunnerInfo, error)
+	EvictTaskRunnerInfo(ctx context.Context, pid int, tid string) error
+	RequeueMessagesFromInProgressQueue(ctx context.Context, inprogressQueue, queue string) ([]string, error)
 
 	// Retries
 	GetAllRetries(ctx context.Context) (*Retries, error)
+
+	// Misc
```
Comment (stefannegrea) [inline storage/storage.go:109]: What do you think having this comment say something like `Storage Server Time`?

File: storage/redis.go [hunk 5]
```diff
@@ -59,52 +59,133 @@ func (r *redisStore) CheckRtt(ctx context.Context) int64 {
...
 	pipe.SAdd(ctx, sidekiqProcessesKey, heartbeat.Identity) // add to the sidekiq processes set without the namespace
 
-	pipe.HSet(ctx, managerIdentity, "beat", heartbeat.Beat.UTC().Unix())
-	pipe.HSet(ctx, managerIdentity, "quiet", heartbeat.Quiet)
-	pipe.HSet(ctx, managerIdentity, "busy", heartbeat.Busy)
-	pipe.HSet(ctx, managerIdentity, "rtt_us", rtt)
-	pipe.HSet(ctx, managerIdentity, "rss", heartbeat.RSS)
-	pipe.HSet(ctx, managerIdentity, "info", heartbeat.Info)
-	pipe.Expire(ctx, managerIdentity, 60*time.Second) // set the TTL of the heartbeat to 60
+	pipe.HMSet(ctx, managerKey, "beat", heartbeat.Beat.UTC().Unix())
```
Comment (stefannegrea) [inline storage/redis.go:75]: Any reason to switch the first command to `HMSet`?

File: fetcher.go (hunk 1 above)
Comment (keitabyte) [reply fetcher.go]: Renamed in fixup: 8ae85e1c6e9047312b29e23e4256a882d7272e08

File: storage/redis.go (hunk 2 above)
Comment (keitabyte) [reply storage/redis.go]: Refactored in fixup: 8ae85e1c6e9047312b29e23e4256a882d7272e08

File: storage/storage.go (hunk 3 above)
Comment (keitabyte) [reply storage/storage.go:77]: While the symmetry is nice, I think being specific about what is required is important. In my opinion it's unclear which properties of the Heartbeat struct are required for the remove to succeed, while the identifier makes it clear. This asymmetry is also common in other APIs (ie: REST has a payload for the POST, but usually just the identifier for the DELETE.)

File: storage/redis.go (hunk 5 above)
Comment (keitabyte) [reply storage/redis.go]: I actually intended to change all the HSets to a single HMSet, but was probably side-tracked mid-change. 😆 
Implemented as HMSet in commit 8ae85e1c6e9047312b29e23e4256a882d7272e08.

File: storage/storage.go [hunk 6]
```diff
@@ -73,9 +92,16 @@ type Store interface {
 	GetAllStats(ctx context.Context, queues []string) (*Stats, error)
 
 	// Heartbeat
+	GetActiveHeartbeatIDs(ctx context.Context) ([]string, error)
+	GetHeartbeat(ctx context.Context, heartbeatID string) (*Heartbeat, error)
```
Comment (stefannegrea) [inline storage/storage.go:96]: How about combining these two into `GetAllHeartbeats()`?

File: storage/storage.go [hunk 7]
```diff
@@ -73,9 +92,16 @@ type Store interface {
...
 	SendHeartbeat(ctx context.Context, heartbeat *Heartbeat) error
-	RemoveHeartbeat(ctx context.Context, heartbeat *Heartbeat) error
+	RemoveHeartbeat(ctx context.Context, heartbeatID string) error
+	HandleExpiredHeartbeatIdentities(ctx context.Context) ([]string, error)
+	HandleExpiredWorkerHeartbeats(ctx context.Context, expireTS int64) ([]*StaleMessageUpdate, error)
```
Comment (stefannegrea) [inline storage/storage.go:100]: How about using the manager heartbeat for this? And not rely on a worker heartbeat?

File: storage/storage.go (hunk 6 above)
Comment (keitabyte) [reply storage/storage.go]: Updated in 5946b8ef.

File: storage/storage.go (hunk 7 above)
Comment (keitabyte) [reply storage/storage.go]: Updated in 75328f44.

File: storage/storage.go [hunk 8]
```diff
@@ -37,17 +37,35 @@ type Retries struct {
...
-	Pid   int
+	Beat            int64  `json:"beat,string"`
+	Quiet           bool   `json:"quiet,string"`
+	Busy            int    `json:"busy,string"`
+	RttUS           int    `json:"rtt_us,string"`
+	RSS             int64  `json:"rss,string"`
+	Info            string `json:"info"`
+	Pid             int    `json:"pid,string"`
+	ManagerPriority int    `json:"manager_priority,string"`
+	ActiveManager   bool   `json:"active_manager,string"`
```
Comment (stefannegrea) [inline storage/storage.go:50]: Any reason to have this as a string?

File: storage/storage.go [hunk 9]
```diff
@@ -37,17 +37,35 @@ type Retries struct {
...
-	RttUS int
-	RSS   int64
-	Info  string
-	Pid   int
+	Beat            int64  `json:"beat,string"`
+	Quiet           bool   `json:"quiet,string"`
+	Busy            int    `json:"busy,string"`
+	RttUS           int    `json:"rtt_us,string"`
+	RSS             int64  `json:"rss,string"`
+	Info            string `json:"info"`
```
Comment (stefannegrea) [inline storage/storage.go:47]: For symmetry, do you mind adding the `string` modifier to all of them?

File: storage/redis.go [hunk 10]
```diff
@@ -57,53 +58,194 @@ func (r *redisStore) CheckRtt(ctx context.Context) int64 {
...
 	if err != nil && err != redis.Nil {
 		return err
 	}
 
 	return nil
 }
 
-func (r *redisStore) RemoveHeartbeat(ctx context.Context, heartbeat *Heartbeat) error {
-	managerIdentity := r.namespace + heartbeat.Identity
+func (r *redisStore) HandleAllExpiredHeartbeats(ctx context.Context, expireTS int64) ([]*StaleMessageUpdate, error) {
```
Comment (stefannegrea) [inline storage/redis.go:163]: Can this code be moved into the manager? And use a combination of `GetAllHeartbeats` and `RequeueMessagesFromInProgressQueue`?

File: storage/redis.go (hunk 2 above)
Comment (stefannegrea) [reply storage/redis.go]: Thank you for updating this.

File: storage/storage.go (hunk 6 above)
Comment (stefannegrea) [reply storage/storage.go]: Thank you for updating this.

File: storage/storage.go (hunk 4 above)
Comment (stefannegrea) [reply storage/storage.go]: Thank you for updating this.

File: storage/storage.go (hunk 9 above)
Comment (keitabyte) [reply storage/storage.go:47]: I attempted doing this, but cannot apply to string type:
https://go.dev/play/p/D7Phyjkwm3f

File: storage/redis.go (hunk 10 above)
Comment (keitabyte) [reply storage/redis.go]: Moved into manager in 5dcbe1fc2191115664209dc63ea3d5a228911580.

File: storage/storage.go (hunk 8 above)
Comment (stefannegrea) [reply storage/storage.go:50]: I see, that makes sense.

File: storage/storage.go (hunk 9 above)
Comment (stefannegrea) [reply storage/storage.go:47]: Ohhhh right! Silly marshaller

File: storage/storage.go [hunk 11]
```diff
@@ -37,17 +37,35 @@ type Retries struct {
...
+
+type WorkerHeartbeat struct {
+	Pid             int    `json:"pid,string"`
+	Tid             string `json:"tid,string"`
+	Queue           string `json:"queue,string"`
+	InProgressQueue string `json:"in_progress_queue,string"`
+	WorkerMsg       string `json:"worker_msg,string"`
+}
+
+type StaleMessageUpdate struct {
```
[the remark this answers is not in the document]
Comment (stefannegrea) [reply storage/storage.go]: Thank you for all these updates.

File: storage/redis.go (hunk 10 above)
Comment (stefannegrea) [reply storage/redis.go]: Awesome work!

File: storage/storage.go (hunk 7 above)
Comment (stefannegrea) [reply storage/storage.go]: Thank you for updating this.

**Review by stefannegrea — APPROVED** [review]

File: storage/redis.go (hunk 5 above)
Comment (skunkworker) [reply storage/redis.go]: @keitabyte this removed some of the functionality where the heartbeat would automatically expire itself. Was that intentional or accidental? It broke compatibility with the Sidekiq Web UI

**Comment by skunkworker** [comment]
@keitabyte @stefannegrea 
It looks like this MR broke compatibility with the Sidekiq Web UI and heartbeats are no longer auto expiring as they were designed to do. The logic has diverged from the Sidekiq 6.* Web UI that it was originally designed around. Was it intentional to make the TTL no longer remove the worker?

**Pull request merged**